csharp/src/Google.Protobuf/CodedInputStream.cs

```diff
@@ -560,27 +783,94 @@ public string ReadString()
             {
                 // Fast path:  We already have the bytes in a contiguous buffer, so
                 //   just copy directly from it.
-                String result = CodedOutputStream.Utf8Encoding.GetString(buffer, bufferPos, length);
+#if NETCOREAPP2_1
+                String result = CodedOutputStream.Utf8Encoding.GetString(immediateBuffer.Slice(bufferPos, length));

```

Comment:
in netcoreapp2.1 the immediateBuffer will always be used?

Reply:
This is a tough question. I started thinking about using buffer if it is available... the best choice depends on the actual Stream implementation. Some has very efficient span version and some are using memory copy... so using buffer if available for .net core 2.1 might be still reasonable